### PR TITLE
PR: Localize strings in JLab

### PR DESCRIPTION
**Pull request by goanpeca**
## Backwards-incompatible changes

To be completed: 

1. File/Edit/Kernel/Run menu extensions that used constants for action/noun/plural/name have now been replaced to functions that take the amount of items so that the pluralization can be correctly handled for different languages.

2. IEditorServices provides now 2 functions `getFactoryService`, `getMimeTypeService`(that accept an optional translator instead of `factoryService` and `typeService`.

```typescript
/**
 * Code editor services.
 */
export interface IEditorServices {
  /**
   * The code editor factory.
   */
  getFactoryService(translator?: ITranslator): IEditorFactoryService;

  /**
   * The editor mime type service.
   */
  getMimeTypeService(translator?: ITranslator): IEditorMimeTypeService;
}

```

3. OptionsMap that use the localizable string as the key. The names of these var will be changed to break (and force) the correct usage.

### CODE REVIEW

File: out.txt [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+packages/translation-extension/package.json
+packages/translation-extension/schema/plugin.json
+packages/translation-extension/src/index.ts
+packages/translation-extension/style/index.css
+packages/translation-extension/tsconfig.json
+packages/translation/LICENSE
+packages/translation/README.md
+packages/translation/babel.config.js
+packages/translation/jest.config.js
+packages/translation/package-lock.json
```
Comment (blink1073) [inline out.txt:310]: I think this file needs to be removed (and it is a huge part of the diff)

**Comment by blink1073** [comment]
I'll be afk for the next 3 hours or so, but will be focusing on getting this done after that.

**Comment by goanpeca** [comment]
>I'll be afk for the next 3 hours or so, but will be focusing on getting this done after that.

As far as I can tell there I sno need to change any API, except for the noun and plural stuff in the menus. So I could open a PR just with that and make the rest of the changes in separate PRs and not in a hurry.

Still let's talk to go over the details. Thanks!

File: Untitled.ipynb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (blink1073) [inline Untitled.ipynb:1]: extra file

File: packages/apputils/src/mainareawidget.ts [hunk 3]
```diff
@@ -254,6 +256,12 @@ export namespace MainAreaWidget {
      * An optional promise for when the content is ready to be revealed.
      */
     reveal?: Promise<any>;
+
+    /**
+     * The Langauage translator.
+     * FIXME: Put this in another place? (Neeeded by the notebook)
```
Comment (blink1073) [inline packages/apputils/src/mainareawidget.ts:262]: I think we discussed that this was fine

File: packages/apputils/src/mainareawidget.ts [hunk 4]
```diff
@@ -254,6 +256,12 @@ export namespace MainAreaWidget {
      * An optional promise for when the content is ready to be revealed.
      */
     reveal?: Promise<any>;
+
+    /**
+     * The Langauage translator.
```
Comment (blink1073) [inline packages/apputils/src/mainareawidget.ts:261]: Spelling

File: packages/apputils/src/sessioncontext.tsx [hunk 5]
```diff
@@ -266,11 +277,12 @@ export namespace ISessionContext {
     readonly autoStartDefault?: boolean;
   }
 
-  export type KernelDisplayStatus =
-    | Kernel.Status
-    | Kernel.ConnectionStatus
-    | 'initializing'
-    | '';
+  export type KernelDisplayStatus = string;
+  // export type KernelDisplayStatus =
```
Comment (blink1073) [inline packages/apputils/src/sessioncontext.tsx:281]: Should this be removed?

File: packages/apputils/src/sessioncontext.tsx [hunk 6]
```diff
@@ -1169,25 +1200,32 @@ export const sessionContextDialogs: ISessionContext.IDialogs = {
   /**
    * Select a kernel for the session.
    */
-  async selectKernel(sessionContext: ISessionContext): Promise<void> {
+  async selectKernel(
+    sessionContext: ISessionContext,
+    translator: ITranslator
```
Comment (blink1073) [inline packages/apputils/src/sessioncontext.tsx:1205]: optional?

File: packages/apputils/src/sessioncontext.tsx [hunk 7]
```diff
@@ -1217,7 +1252,13 @@ export const sessionContextDialogs: ISessionContext.IDialogs = {
    * If there is no kernel, we start a kernel with the last run
    * kernel name and resolves with `true`.
    */
-  async restart(sessionContext: ISessionContext): Promise<boolean> {
+  async restart(
+    sessionContext: ISessionContext,
+    translator: ITranslator
```
Comment (blink1073) [inline packages/apputils/src/sessioncontext.tsx:1257]: optional?

File: packages/celltags/src/addwidget.ts [hunk 8]
```diff
@@ -9,8 +14,10 @@ export class AddWidget extends Widget {
   /**
    * Construct a new tag widget.
    */
-  constructor() {
+  constructor(translator: ITranslator) {
```
Comment (blink1073) [inline packages/celltags/src/addwidget.ts:17]: optional?

File: packages/celltags/src/tool.ts [hunk 9]
```diff
@@ -15,9 +20,15 @@ export class TagTool extends NotebookTools.Tool {
    *
    * @param tracker - The notebook tracker.
    */
-  constructor(tracker: INotebookTracker, app: JupyterFrontEnd) {
+  constructor(
+    tracker: INotebookTracker,
+    app: JupyterFrontEnd,
+    translator: ITranslator
```
Comment (blink1073) [inline packages/celltags/src/tool.ts:26]: Optional?

File: packages/codeeditor/src/tokens.ts [hunk 10]
```diff
@@ -30,3 +32,7 @@ export interface IEditorServices {
    */
   readonly mimeTypeService: IEditorMimeTypeService;
 }
+
+export interface IEditorServices2 {
```
Comment (blink1073) [inline packages/codeeditor/src/tokens.ts:36]: was this intentional?

File: packages/console-extension/src/index.ts [hunk 11]
```diff
@@ -254,13 +259,15 @@ async function activateConsole(
    * Create a console for a given path.
    */
   async function createConsole(options: ICreateOptions): Promise<ConsolePanel> {
+    console.log('async function createConsole(', translator);
```
Comment (blink1073) [inline packages/console-extension/src/index.ts:262]: console log

File: packages/apputils/src/sessioncontext.tsx (hunk 5 above)
Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:281]: Ah yes I forgot... so the Kernel status was hardcoding the status actual text in the type... .so we cannot localize it. So I just made kernel display status be a string... any other way to handle this?

File: packages/codeeditor/src/tokens.ts (hunk 10 above)
Comment (goanpeca) [reply packages/codeeditor/src/tokens.ts:36]: Oops!

File: packages/apputils/src/sessioncontext.tsx (hunk 5 above)
Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:281]: Fixed.

File: packages/apputils/src/sessioncontext.tsx (hunk 6 above)
Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:1205]: Fixed.

File: packages/apputils/src/sessioncontext.tsx (hunk 7 above)
Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:1257]: Fixed.

File: packages/celltags/src/addwidget.ts (hunk 8 above)
Comment (goanpeca) [reply packages/celltags/src/addwidget.ts:17]: Fixed. (going over a search of all translators...)

File: packages/celltags/src/tool.ts (hunk 9 above)
Comment (goanpeca) [reply packages/celltags/src/tool.ts:26]: Fixed.

File: packages/codeeditor/src/tokens.ts (hunk 10 above)
Comment (goanpeca) [reply packages/codeeditor/src/tokens.ts:36]: Ahhh yes, this was the thing I wanted to ask about. The factories are currently constants, so they should probably be methods that take the optional translator so we can propagate them down the factory.

Thoughts?

```typescript
/**
 * Code editor services.
 */
export interface IEditorServices {
  /**
   * The code editor factory.
   */
  readonly factoryService: IEditorFactoryService;

  /**
   * The editor mime type service.
   */
  readonly mimeTypeService: IEditorMimeTypeService;
}
```

to

```typescript

/**
 * Code editor services.
 */
export interface IEditorServices {
  /**
   * The code editor factory.
   */
  readonly getFactoryService(translator?: ITranslator): IEditorFactoryService;

  /**
   * The editor mime type service.
   */
  readonly getMimeTypeService(translator?: ITranslator): IEditorMimeTypeService;
}
```

File: packages/console-extension/src/index.ts (hunk 11 above)
Comment (goanpeca) [reply packages/console-extension/src/index.ts:262]: Fixed

File: packages/apputils/src/mainareawidget.ts (hunk 4 above)
Comment (goanpeca) [reply packages/apputils/src/mainareawidget.ts:261]: Fixed.

File: packages/apputils/src/mainareawidget.ts (hunk 3 above)
Comment (goanpeca) [reply packages/apputils/src/mainareawidget.ts:262]: Fixed.

File: Untitled.ipynb (hunk 2 above)
Comment (goanpeca) [reply Untitled.ipynb:1]: Removed.

File: packages/apputils/src/sessioncontext.tsx (hunk 5 above)
Comment (blink1073) [reply packages/apputils/src/sessioncontext.tsx:281]: It comes from the server but there are a finite number of them, so we could translate them all.

File: packages/codeeditor/src/tokens.ts (hunk 10 above)
Comment (blink1073) [reply packages/codeeditor/src/tokens.ts:36]: Yep, sounds reasonabl

File: packages/fileeditor-extension/src/commands.ts [hunk 12]
```diff
@@ -879,60 +903,76 @@ export namespace Commands {
...
 
-    addChangeFontSizeCommandsToPalette(palette);
+    addChangeFontSizeCommandsToPalette(palette, trans);
   }
 
   /**
    * The category for File Editor palette commands for use in addToPalette functions
    */
-  const paletteCategory = 'Text Editor';
+  // const paletteCategory = trans.__('Text Editor');
```
Comment (blink1073) [inline packages/fileeditor-extension/src/commands.ts:962]: wip?

File: packages/translation/src/gettext.ts [hunk 13]
```diff
@@ -266,8 +266,12 @@ class Gettext {
    * Shorthand for gettext.
    *
    * @param msgid - The singular string to translate.
+<<<<<<< HEAD
```
Comment (blink1073) [inline packages/translation/src/gettext.ts:269]: Merge markers throughout this file

**Comment by goanpeca** [comment]
>I didn't look closely at the FIXME comments in this pass

Looking into them after I make the next push (moving the factories changed a lot, so that is also a breaking change I guess... since It is also being used in the exampleS)

File: packages/translation/src/gettext.ts (hunk 13 above)
Comment (goanpeca) [reply packages/translation/src/gettext.ts:269]: Fixed

File: packages/fileeditor-extension/src/commands.ts (hunk 12 above)
Comment (goanpeca) [reply packages/fileeditor-extension/src/commands.ts:962]: Fixed

File: packages/apputils/src/sessioncontext.tsx (hunk 5 above)
Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:281]: Or should I turn this type into an object with a status (that can remain the same as it is) and a displayStatus that is the translated string?

Comment (blink1073) [reply packages/apputils/src/sessioncontext.tsx:281]: I like displayStatus

Comment (goanpeca) [reply packages/apputils/src/sessioncontext.tsx:281]: So to double check. Initialize this object like?

`{[status: KernelDisplayStatus]: string}`  somewhere and the use it to get the localized display status where needed?

Comment (blink1073) [reply packages/apputils/src/sessioncontext.tsx:281]: Hmm, I think `getKernelDisplayStatus()` returns a translated `string` now that you mention it.

**Comment by blink1073** [comment]
I'm addressing some failures in master in #8810

**Comment by goanpeca** [comment]
@blink1073 question. Running the integrity changes I get:

```
jlpm run integrity
yarn run v1.21.1
$ node scripts/ensure-buildutils.js && node buildutils/lib/ensure-repo.js
npm WARN lifecycle The node binary used for scripts is /var/folders/c5/l0n5fsv96gg2218vchtvljdc0000gn/T/yarn--1597153298891-0.8590735466852994/node but npm is using /Users/goanpeca/miniconda3/envs/jlab/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> @jupyterlab/buildutils@3.0.0-alpha.8 build /Users/goanpeca/Dropbox (Personal)/develop/quansight/jupyterlab/buildutils
> tsc

{
  "@jupyterlab/console": [
    "Unused dependency: ts-jest@^25.2.1: remove or add to list of known unused dependencies for this package",
    "Unused dependency: typedoc@^0.17.7: remove or add to list of known unused dependencies for this package",
    "Unused dependency: typescript@~3.9.2: remove or add to list of known unused dependencies for this package"
  ]
}
> jlpm install
[1/4] 🔍  Resolving packages...
success Already up-to-date.
$ node scripts/ensure-buildutils.js
npm WARN lifecycle The node binary used for scripts is /var/folders/c5/l0n5fsv96gg2218vchtvljdc0000gn/T/yarn--1597153310587-0.4112658597871257/node but npm is using /Users/goanpeca/miniconda3/envs/jlab/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was executed with.

> @jupyterlab/buildutils@3.0.0-alpha.8 build /Users/goanpeca/Dropbox (Personal)/develop/quansight/jupyterlab/buildutils
> tsc



Made integrity changes!
Please commit the changes by running:
git commit -a -m "Package integrity updates"
```

But nothing actually changed to commit. 


Also for the console tests I get

```
~/work/jupyterlab/jupyterlab/packages/console ~/work/jupyterlab/jupyterlab
yarn run v1.21.1
$ tsc --build tsconfig.test.json
##[error]test/foreign.spec.ts(16,48): error TS2307: Cannot find module '@jupyterlab/testutils' or its corresponding type declarations.
##[error]test/foreign.spec.ts(18,23): error TS2307: Cannot find module '@jupyterlab/testutils/lib/mock' or its corresponding type declarations.
##[error]test/history.spec.ts(14,55): error TS2307: Cannot find module '@jupyterlab/testutils' or its corresponding type declarations.
##[error]test/utils.ts(6,35): error TS2307: Cannot find module '@jupyterlab/testutils' or its corresponding type declarations.
##[error]test/panel.spec.ts(12,31): error TS2307: Cannot find module '@jupyterlab/testutils' or its corresponding type declarations.
##[error]test/panel.spec.ts(14,23): error TS2307: Cannot find module '@jupyterlab/testutils/lib/mock' or its corresponding type declarations.
##[error]test/widget.spec.ts(21,51): error TS2307: Cannot find module '@jupyterlab/testutils' or its corresponding type declarations.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
##[error]Process completed with exit code 2.
```

If I manually add `../../testutils` to `tsconfig.test.json`  it gets removed on the build / or thee lint step (did not check which one)

**Comment by blink1073** [comment]
Not sure yet, let me finish #8810

**Comment by goanpeca** [comment]
I also changed the `FIXME:` comments to `FIXME-TRANS:` so we can tackled them faster.

File: .github/workflows/linuxtests.yml [hunk 14]
```diff
@@ -27,7 +27,7 @@ jobs:
           - group: examples
             python: 3.5
       fail-fast: false
-    timeout-minutes: 120
+    timeout-minutes: 15 # 120
```
Comment (blink1073) [inline .github/workflows/linuxtests.yml:30]: There are a couple tests that go over 20 minutes

**Comment by blink1073** [comment]
To fix examples:

- add `@jupyterlab/translation-extension` to `examples/app/index.js`
- add `@jupyterlab/translation-extension` and `@jupyterlab/translation` deps in `examples/app/package.json`
- add `@jupyterlab/translation-extension` and `@jupyterlab/translation` deps in `examples/federated/core_package/package.json`
- also put both them in `resolutions` and put the extension in `extensions` in that same `core_package`

**Comment by afshin** [comment]
I think this is causing the file browser plugin to fail:
```diff
diff --git a/packages/filebrowser-extension/src/index.ts b/packages/filebrowser-extension/src/index.ts
index 90f662256..190e35c96 100644
--- a/packages/filebrowser-extension/src/index.ts
+++ b/packages/filebrowser-extension/src/index.ts
@@ -158,7 +158,7 @@ const factory: JupyterFrontEndPlugin<IFileBrowserFactory> = {
   activate: activateFactory,
   id: '@jupyterlab/filebrowser-extension:factory',
   provides: IFileBrowserFactory,
-  requires: [ILabShell, IDocumentManager, IDocumentManager, ITranslator],
+  requires: [ILabShell, IDocumentManager, ITranslator],
   optional: [IStateDB, IRouter, JupyterFrontEnd.ITreeResolver]
 };
```

**Comment by goanpeca** [comment]
Thanks for the help @afshin ! I will add a check for number of parameters on requires and also check if a same plugin has been registered before. That very stupid error on my side was not providing the right feedback :-p

**Pull request merged**